primap2/csg/_compose.py

```diff
@@ -77,52 +78,52 @@ def compose(
         describe the processing steps done for each timeseries.
     """
     result_das = {}
+    input_data = input_data.pr.dequantify()
 
     if progress_bar is None:
         entity_iterator = input_data
     else:
         entity_iterator = progress_bar(input_data)
+
+    priority_dimensions = priority_definition.priority_dimensions
+    priority_definition.check_dimensions()
+
     for entity in entity_iterator:
         if progress_bar is not None:
-            entity_iterator.set_postfix_str(entity)
-        input_da = input_data[entity].pr.dequantify()
+            entity_iterator.set_postfix_str(str(entity))
+
+        input_da = input_data[entity]
+
         # all dimensions are either time, priority selection dimensions, or need to
         # be iterated over
         group_by_dimensions = tuple(
             dim
             for dim in input_da.dims
-            if dim != "time" and dim not in priority_definition.priority_dimensions
-        )
-        # pre-allocate result arrays which will be filled timeseries-by-timeseries
-        result_dimensions = ["time", *group_by_dimensions]
-        result_das[entity] = xr.DataArray(
-            data=np.nan,
-            dims=result_dimensions,
-            coords={
-                dim: input_da.coords[dim]
-                for dim in input_da.coords
-                if dim not in priority_definition.priority_dimensions
-            },
-            attrs=input_da.attrs,
+            if dim != "time" and dim not in priority_dimensions
         )
-        result_das[f"Processing of {entity}"] = xr.DataArray(
-            data=np.empty(
-                [len(input_da.coords[dim]) for dim in group_by_dimensions], dtype=object
-            ),
-            dims=group_by_dimensions,
-            coords=[input_da.coords[dim] for dim in group_by_dimensions],
+
+        (
+            result_das[entity],
+            result_das[f"Processing of {entity}"],
+        ) = preallocate_result_arrays(
+            input_da=input_da,
+            group_by_dimensions=group_by_dimensions,
+            priority_dimensions=priority_dimensions,
         )
+
         number_of_timeseries = math.prod(
             len(input_da[dim]) for dim in group_by_dimensions
         )
+
         if progress_bar is None:
             pbar = None
         else:
             pbar = progress_bar(total=number_of_timeseries, unit="ts", unit_scale=True)
+
         iterate_next_fixed_dimension(
             input_da=input_da,
-            priority_definition=priority_definition,
-            strategy_definition=strategy_definition,
+            priority_definition=priority_definition.limit("entity", entity),
+            strategy_definition=strategy_definition.limit("entity", entity),
```

Comment:
it was surprisingly easy to enable filtering by entity, these two lines are all that's needed.

Reply:
Nice. didn't expect that but in hindsight I could have expected it as we're working on a per entity basis anyway, so no need for additional filtering of entities.